### DISCUSSION TITLE
FR: "limit to subfolders" for Assets

### DISCUSSION THREAD

**Discussion by adrienne** [craftcms/cms#9070] [Ideas] [answered]
I mentioned this on the Discord and Brandon said to write up a FR over here, so I am. 

Currently there are two mutually exclusive options for the location/saving permissions of Assets fields. Quoting from the documentation:
> Restrict uploads to a single folder? – Whether file uploads/relations should be constrained to a single folder.
>
> If enabled, the following setting will be visible:
> - Upload Location – The location that files dragged directly onto the field should be saved in.
>
> If disabled, the following settings will be visible:
> 
> - Sources – Which asset volumes (or other asset index sources) the field should be able to relate assets from.
> - Default Upload Location – The default location that files dragged directly onto the field should be saved in.

What i would really like is a third option, "Restrict to folder tree" or something, which would work as follows:

- it would, like "Restrict uploads to a single folder", allow a single selectable volume and a configurable subfolder, and prevent viewing or uploading files *outside* of itself 
- *unlike* "Restrict uploads...", it would allow subfolders *within* itself
    - So if i set it to, for instance, `my/folder`, i would be able to access `my/folder` but also `my/folder/sub1` and `my/folder/sub2`. And importantly:
- It would still enable the "Default Upload Location" setting, such that subfolders could be automatically created within the root folder
    - So if `my/folder` was the "Restrict to..." setting, and then my Default Upload location was `{{ postDate|format('Y') }}`, images I uploaded would end up in `my/folder/2021`
- Like both of the above settings, it would allow Twig configuration (so that we could use it to restrict the subfolder hierarchy to `{{ site.handle }}`, for instance)

I understand this is a little complex to describe, but it seems like the *easiest* possible solution for making Assets fields really cross-site compatible without everyone getting confused about where files are. (And I can think of many other use cases, as well!)

**Comment by brandonkelly** [comment]
Alternatively, we could add an “Allow subfolders” sub-setting of the “Restrict” option.

**Reply by adrienne** [reply]
Would also work, @brandonkelly, but you'd ideally still need some kind of setting to configure the subfolders. Like, here's my use case:

- I want the field _as a whole_ to be restricted to: `{{ currentSite.handle }}/` so that nobody can _ever_ access an asset outside the current site using the "select a file" button inside the field
- And then i also want any uploads to _default_, within that folder, to subfolders `{{ postDate|format('Y') }}/{{ postDate|format('M') }}/{{ username }}`

**Reply by AndrewMac** [reply]
Certainly something I’d really appreciate, and I totally agree with the need to have the default subfolder option

**Reply by andris-sevcenko** [reply]
@adrienne To follow up on your use case - there are some challenges to overcome for a hybrid solution and perhaps you have an input on how you would see it work. Here's the thing.

(1) The default upload location only kicks in when dropping a file on the assets field. On opening the assets modal for the field, only volume restrictions apply. (I'm disregarding front-end upload scenarios because that's not entirely relevant here)
(2) The single upload folder setting opens the assets modal without displaying any folders.
(3) Allowing a subtree would require displaying the folder tree, obviously.

**Problem (A)**
Both (1) and (2) conveniently sidestep the problem of displaying a folder that might not exist yet or a folder that's name or location in the tree cannot be figured out yet. In the example you posted, it's impossible to know both the name and the location of the folder until there is a post date set on the entry you are editing. So, you'd need to display _both_ a folder tree and some _magic_ folder, that will change its path, once it can be resolved.

**Problem (B)**
Currently, when an element gets saved, the assets fields on the element look at their settings - if the single upload folder setting is enabled, the assets field can be reasonably sure, that, if the folder name changes based on element changes, the asset should be moved to the new folder. E.g. if the post date is changed on the entry, the asset should be moved to a new folder to reflect that.

In the case of having a hybrid approach, Craft can't make that assumption anymore, since each asset inside the field has to be treated differently. This means that, most likely, once a folder name can be resolved, the asset won't be moved again.


I'm not saying those challenges are unsolvable, it just makes things a bit more complicated than it seems at a glance.

**Reply by adrienne** [reply]
@andris-sevcenko Hm, yeah, I can see that. Let me consider!

**Comment by brandonkelly** [comment, marked as answer]
Done for Craft 4! (#10080)

- Renamed **Restrict assets to a single folder** to **Restrict assets to a single location**
- When that’s selected, a new **Allow subfolders** checkbox is shown below it
- When that’s selected, a **Default Upload Location** setting is shown (similar to the one you’d get when **Restrict assets to a single location** is **not** selected, except this one doesn’t have a source/volume menu; only a subpath.

Here’s what the settings look like when assets are restricted:

<img width="645" alt="Screenshot of a portion of an Assets field’s settings, where the “Restrict assets to a single location” checkbox is checked; the “Asset Location” setting is set to “Site Assets → {site.handle}”, the “Allow subfolders” setting is checked; and the “Default Upload Location” setting is set to “{slug}”." src="https://user-images.githubusercontent.com/47792/144729106-bbd79468-e5cc-4d38-b043-d10562206dda.png">

And when they’re not:

<img width="646" alt="Screenshot of a portion of an Assets field’s settings, where the “Restrict assets to a single location” checkbox is checked; the “Sources” setting is set to a single “Site Assets” volume; and the “Default Upload Location” setting is set to “Site Assets → {site.handle}”." src="https://user-images.githubusercontent.com/47792/144729191-d648ae2c-d484-4aa3-8359-4ca7d76f7534.png">

**Reply by adrienne** [reply]
omg @brandonkelly you guys really are the best.
